src/java.base/share/classes/jdk/internal/jimage/ImageReader.java

```diff
@@ -92,146 +125,113 @@ private void requireOpen() {
-        return reader.findNode("/modules")
-                     .getChildren()
-                     .stream()
-                     .map(Node::getNameString)
-                     .map(s -> s.substring(off, s.length()))
-                     .toArray(String[]::new);
-    }
-
-    public long[] getAttributes(int offset) {
-        requireOpen();
-        return reader.getAttributes(offset);
-    }
-
-    public String getString(int offset) {
-        requireOpen();
-        return reader.getString(offset);
-    }
-
-    public byte[] getResource(String name) {
-        requireOpen();
-        return reader.getResource(name);
-    }
-
-    public byte[] getResource(ImageLocation loc) {
-        requireOpen();
-        return reader.getResource(loc);
-    }
-
-    public ByteBuffer getResourceBuffer(ImageLocation loc) {
-        requireOpen();
-        return reader.getResourceBuffer(loc);
-    }
-
-    public InputStream getResourceStream(ImageLocation loc) {
-        requireOpen();
-        return reader.getResourceStream(loc);
+        if (!node.isResource()) {
+            throw new IllegalStateException("Not a resource node: " + node);
+        }
+        return reader.getResourceBuffer(node.getLocation());
     }
 
     private static final class SharedImageReader extends BasicImageReader {
-        static final int SIZE_OF_OFFSET = Integer.BYTES;
-
-        static final Map<Path, SharedImageReader> OPEN_FILES = new HashMap<>();
+        private static final Map<Path, SharedImageReader> OPEN_FILES = new HashMap<>();
+        private static final String MODULES_ROOT = "/modules";
+        private static final String PACKAGES_ROOT = "/packages";
+        // There are >30,000 nodes in a complete jimage tree, and even relatively
+        // common tasks (e.g. starting up javac) load somewhere in the region of
+        // 1000 classes. Thus, an initial capacity of 2000 is a reasonable guess.
+        private static final int INITIAL_NODE_CACHE_CAPACITY = 2000;
 
         // List of openers for this shared image.
-        final Set<ImageReader> openers;
+        private final Set<ImageReader> openers = new HashSet<>();
 
-        // attributes of the .jimage file. jimage file does not contain
+        // Attributes of the .jimage file. The jimage file does not contain
         // attributes for the individual resources (yet). We use attributes
         // of the jimage file itself (creation, modification, access times).
-        // Iniitalized lazily, see {@link #imageFileAttributes()}.
-        BasicFileAttributes imageFileAttributes;
+        private final BasicFileAttributes imageFileAttributes;
 
-        // directory management implementation
-        final HashMap<String, Node> nodes;
-        volatile Directory rootDir;
-
-        Directory packagesDir;
-        Directory modulesDir;
+        // Cache of all user visible nodes, guarded by synchronizing 'this' instance.
+        private final HashMap<String, Node> nodes;
```

Comment:
Pre-existing, but since we are cleaning up some of this code, perhaps we can make this declaration a `Map` instead of a `HashMap`?